### PR TITLE
Total refactor of Link

### PR DESCRIPTION
**Pull request by butlermatt**
Link was initially one massive function. Break it up into components.

### CODE REVIEW

File: lib/node_manager.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  Future<Null> valueSubscribe(ServerRequest sr, WebSocket socket);
+}
+
+class ServerRequest {
+  String path;
+  HttpRequest request;
+  HttpResponse get response => request.response;
+  String get method => request.method;
+  bool isHtml = false;
+  bool get binary => request.uri.queryParameters.containsKey('binary');
```
Comment (Pacane) [inline lib/node_manager.dart:21]: I'd keep the same naming convention here, either with `isXXX` or `hasXXX`, and personally I prefer `isBinary` over `binary`.

File: lib/nodes/create_value.dart [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  CreateNode(String path, this.link): super(path);
+
+  @override
+  Future<Map<String, dynamic>> onInvoke(Map<String, String> params) async {
+    var ret = {_success: true, _message: 'Success!'};
+    var name = params["name"];
+
+    var parent = new Path(path).parent;
+    provider.addNode("${parent.path}/${name}", RestNode.def());
+    link.saveAsync();
```
Comment (Pacane) [inline lib/nodes/create_value.dart:107]: don't you want to await those `saveAsync`s? if so, you might want to change that one and the others.

**Review by Pacane — APPROVED** [review]
Can you dartfmt after merging this?

File: lib/nodes/create_value.dart (hunk 2 above)
Comment (butlermatt) [reply lib/nodes/create_value.dart:107]: No, don't need to await them as they don't return a value I need to complete. They're 'fire and forget' so to speak.

File: lib/node_manager.dart (hunk 1 above)
Comment (butlermatt) [reply lib/node_manager.dart:21]: done. (and same with hasChildValues just below this)

**Pull request merged**